### ISSUE TITLE
Map过低

### ISSUE THREAD

**Issue by jszqf** [yz-wang/TogetherNet#2]
您好！我按照您的论文设置训练了一下，这是我训练了10、20epoch的结果，效果似乎有点糟糕，请问这是正常现象吗？损失函数也没重合，似乎遇到了过拟合现象？
![epoch_loss](https://user-images.githubusercontent.com/51877252/217256219-d89dc378-d78a-4d16-be89-44c9a240e6ac.png)
![mAP](https://user-images.githubusercontent.com/51877252/217256015-b7b3eecb-b008-4d70-8987-53c6d895e339.png)
![mAP20](https://user-images.githubusercontent.com/51877252/217256037-eaffdc3b-f1b7-4ef2-8b73-3f6db29ab10c.png)

**Comment by jszqf** [comment]
这是100个epoch的结果，似乎不容乐观
![epoch_loss](https://user-images.githubusercontent.com/51877252/217416921-08b4c937-3373-44bb-9ebf-9095a97cda35.png)
![Snipaste_2023-02-08_10-52-51](https://user-images.githubusercontent.com/51877252/217416925-d91224f0-543e-4086-ab6d-fd3d286494ec.png)

**Comment by yz-wang** [comment]
可能是你数据集的label没处理好，或动了啥参数，因为原始YOLOXs就能跑到80左右的mAP了

**Comment by ssmemg** [comment]
您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下：
.../../../xxx.jpg <x , y , w , h , cls_id>
Example:
/home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0
![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png)
训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

**Comment by yz-wang** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x , y , w , h , cls_id> Example: /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

邮件已回复

**Comment by ahxiaofengzheng** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x ， y ， w ， h ， cls_id> 示例： /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208，263，333，474，0 ![图像](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

请问您的数据集时如何划分的，代码中需要修改2007_train_fog.txt、2007_val_fog.txt、2007_train.txt、2007_val.tx这四个txt文件，您是如何修改的，能否展示一下您修改的文件。

**Comment by KA1K1** [comment]
> > 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x ， y ， w ， h ， cls_id> 示例： /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208，263，333，474，0 ![图像](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢
> 
> 请问您的数据集时如何划分的，代码中需要修改2007_train_fog.txt、2007_val_fog.txt、2007_train.txt、2007_val.tx这四个txt文件，您是如何修改的，能否展示一下您修改的文件。

同问，解决了吗？望赐教

**Comment by AmeryXiong** [comment]
![image](https://github.com/yz-wang/TogetherNet/assets/102467128/9067a2be-6e01-476d-8e20-a07ab4df8814)
请问为什么训练出来，map值比较低，达不到论文的结果。

**Comment by AmeryXiong** [comment]
![image](https://github.com/yz-wang/TogetherNet/assets/102467128/91188f4e-9ce2-4fc9-b4c4-6da8e7384be9)
请问为什么voc训练出来的map值达不到论文的结果。

**Comment by yz-wang** [comment]
> ![image](https://user-images.githubusercontent.com/102467128/254865990-91188f4e-9ce2-4fc9-b4c4-6da8e7384be9.png) 请问为什么voc训练出来的map值达不到论文的结果。

建议在预训练好的YOLOXS模型下重新训练，比如COCO数据集，利用大数据集帮助新模型快速收敛和提高AP值

**Comment by YHWH666** [comment]
请问我在使用voc-fog数据集训练验证时mAP仅有15%左右，并且在使用yolox_s.pth预训练权重进行训练时会报错，具体如下图所示
![训练报错](https://github.com/yz-wang/TogetherNet/assets/73100535/29df6183-e855-41a9-ad76-b49e6efc2b29)

**Comment by YHWH666** [comment]
我在使用voc-fog数据集训练验证时mAP仅有15%左右，并且在使用yolox_s.pth预训练权重进行训练时会报错，具体如下图所示，请问这个是什么原因导致的呢？
![训练报错](https://github.com/yz-wang/TogetherNet/assets/73100535/1d0fe629-9133-4458-9258-da17d4ec9834)

**Comment by YHWH666** [comment]
请问可以提供一下那四个txt标签文件吗？不确定目前生成的标签是否正确，谢谢！

**Comment by bksheng** [comment]
> > 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x , y , w , h , cls_id> Example: /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢
> 
> 邮件已回复

作者您好，可以发一下这个label格式的示例吗？

**Comment by bksheng** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x , y , w , h , cls_id> Example: /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

你好，请问你后来使用的格式是什么样的？

**Comment by bksheng** [comment]
> ![image](https://private-user-images.githubusercontent.com/102467128/254865990-91188f4e-9ce2-4fc9-b4c4-6da8e7384be9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwNzE2MTksIm5iZiI6MTcwMTA3MTMxOSwicGF0aCI6Ii8xMDI0NjcxMjgvMjU0ODY1OTkwLTkxMTg4ZjRlLTljZTItNGZjOS1iNGM0LTZkYThlNzM4NGJlOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTEyN1QwNzQ4MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZGJjOWQzNGMxZDYzMDQxZjU1MjRjZDY0YmQ1OTEzMWVjYmZlY2I3MWJlMDZhYWExY2VkOWJiZjZmMGFlMGU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.llr_8Ji4_TsJndOijcwKoUZsh0C7St3b995UkfQ8cI4) 请问为什么voc训练出来的map值达不到论文的结果。

你好，请问你是用哪个数据集训练的？ 可以说下你的train.txt是什么样的吗？

**Comment by llljjww** [comment]
您好，我也遇到了MAP过低的问题，也没有改动过参数，生成的TXT文件如下
![图片1](https://github.com/yz-wang/TogetherNet/assets/152797205/ca2ad2a3-4d0e-4e93-921b-9cbe642ffc03)
请问生成的格式正确吗？谢谢您

**Comment by buyi6666** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x ， y ， w ， h ， cls_id> 示例： /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![图像](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

您好！我也遇到了这个问题，请问您解决了吗？如果有消息还麻烦您帮帮我！万分感谢~~~

**Comment by buyi6666** [comment]
您好！谢谢您的贡献！❀❀❀

我在复现代码的过程中，也遇到了训练结果map值低到离谱的情况。猜测可能是label格式不对，请问可以把您的格式告诉我吗？
![image](https://github.com/user-attachments/assets/aab7247b-8f5a-4014-a094-e3dcc2dd2c44)

![image](https://github.com/user-attachments/assets/d5529a46-fc30-42be-9cc0-0ded4f77746a)

**Comment by liu-yv-lin** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x , y , w , h , cls_id> Example: /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

你好，我想问一下，togethernet代码中怎么生成yolo label格式呢，我现在运行不了代码

**Comment by liu-yv-lin** [comment]
> 您好！我按照您的论文设置训练了一下，这是我训练了10、20epoch的结果，效果似乎有点糟糕，请问这是正常现象吗？损失函数也没重合，似乎遇到了过拟合现象？ ![epoch_loss](https://user-images.githubusercontent.com/51877252/217256219-d89dc378-d78a-4d16-be89-44c9a240e6ac.png) ![mAP](https://user-images.githubusercontent.com/51877252/217256015-b7b3eecb-b008-4d70-8987-53c6d895e339.png) ![mAP20](https://user-images.githubusercontent.com/51877252/217256037-eaffdc3b-f1b7-4ef2-8b73-3f6db29ab10c.png)

你好，请问一下，怎么能生成yolo格式呢

**Comment by liu-yv-lin** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x , y , w , h , cls_id> Example: /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

你好，请问一下这里面怎么生成yolo的格式呢，谢谢了

**Comment by liu-yv-lin** [comment]
> 您好，首先感谢你的贡献，我遇到了同样的问题，参数没做任何改动，我生成的YOLO label格式如下： .../../../xxx.jpg <x , y , w , h , cls_id> Example: /home/mengsensen/code/TogetherNet/VOCdevkit/VOC2007/VOC2007-FOG/2011_004672.jpg 208,263,333,474,0 ![image](https://user-images.githubusercontent.com/115381311/219040033-5e562e2c-e68b-4a35-8b0b-a5e92bccde22.png) 训练出来的结果和上述基本一致，麻烦问一下我生成的转化标签是否正确，可否麻烦您提供一下您生成的train.txt和val.txt文件或者提供一个示例？谢谢

你好，请问一下怎么生成YOLO格式，作者将他生成的train.txt提供给你了吗？

**Comment by yehengkang** [comment]
> ![image](https://github.com/yz-wang/TogetherNet/assets/102467128/91188f4e-9ce2-4fc9-b4c4-6da8e7384be9) 请问为什么voc训练出来的map值达不到论文的结果。

这个不是因为预训练模型的加载问题。我在复现过程中发现了和你一样的问题，并成功解决。
之前有个issue里面说 多卡训练和数据分割无法共存 ，这个issue给了我启发（他那个问题也不是真的无法共存，而是他没有找到合适的方法）：
nets.yolo.py里的forward部分：
    def forward(self, input):
        if self.training:
            input, clear_x = input.split((4, 4), dim=0)  # split haze and clear images (Batchsize, Batchsize)
        out_features            = self.backbone.forward(input)
        [feat1, feat2, feat3]   = [out_features[f] for f in self.in_features]

可以看出他这里把拼接之后的8个通道分隔开了，然后把前四个batch作为四个模糊图像，后四个作为清晰图像。
这是问题发生的源头。

然后找到这个拼接图是在哪来的：是在utils.utils_fit.py这里。
        with torch.no_grad():
            if cuda:
                images  = torch.from_numpy(images).type(torch.FloatTensor).cuda(local_rank)
                targets = [torch.from_numpy(ann).type(torch.FloatTensor).cuda(local_rank) for ann in targets]
                clearimgs = torch.from_numpy(clearimgs).type(torch.FloatTensor).cuda(local_rank)
                hazy_and_clear = torch.cat([images, clearimgs], dim=0).cuda()

这个  hazy_and_clear   就是后续传到forward里面的input。
本来按理说，先组合，再分开，应该没什么问题。
但是当使用多卡训练的时候，由于DP的机制，输入数据会分发到不同的卡里面，就造成了这个问题。
举个例子，如果这里batch_size 是16，用四个卡训练。
那hazy_and_clear 的batch应该是 16 + 16 = 32，然后这32会分到四个卡里面，每个卡是8。看起来正好input.split((4, 4), dim=0)。

但是！DP的机制的分割到不同卡里是按照顺序来的，也就是说，前八个（全是模糊图像）分到了第一张卡；再八个（全是模糊图像）分到了第二张卡；再八个（全是清晰图像！）分到了第三张卡；最后八个（全是清晰图像！）分到了最后一张卡。

这就会导致训练的时候，后面两张卡直接输入的都是清晰图像。

这才是导致这个mAP的根本原因。

解决方法也很简单，观察发现，nets.yolo.py里的forward部分根本没有用到clear_x ，所以，直接在utils.utils_fit.py这里，把拼接的这一行删去。然后对应的，把
if self.training:
            input, clear_x = input.split((4, 4), dim=0)  # split haze and clear images (Batchsize, Batchsize)
这两行也删去就行了。（而且删去了拼接的batch加倍，这里正好也删去了分割的batch减半），所以不用调整其他地方的代码也是可以正常训练的。（对了，因为删去了hazy_and_clear这个变量，后续outputs = model_train(hazy_and_clear)这里的输入要改成outputs = model_train(images)，综上，该这三点就可以了）

然后训练得到的最后结果是 mAP 79 左右，跟这个“issue里面说 多卡训练和数据分割无法共存 ”的结果差不多。也打不到原文85那么高。但是这应该是其他原因了。还没找到
